linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -4212,27 +4222,65 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
     } else if (mutable) {
       withDedicatedVar(RefinedType(declaredType))
     } else {
-      val refinedType = value.tpe
+      def computeRefinedType(): RefinedType = bindingName match {
+        case _ if value.tpe.isExact || declaredType == AnyType =>
+          value.tpe
+        case _: Binding.Local =>
+          value.tpe
+        case Binding.This =>
+          /* When binding to `this`, if the declared type is not `AnyType`,
+           * we are in a situation where
+           * a) we know the value must be non-null, and
+           * b) the declaredType may bring more precise information than
+           *    value.tpe.base (typically when inlining a polymorphic method
+           *    that ends up having only one target in a subclass).
+           * We can refine the type here based on that knowledge
+           */
+          val improvedBaseType =
+            if (isSubtype(value.tpe.base, declaredType)) value.tpe.base
+            else declaredType
+          val isExact = false // We catch the case value.tpe.isExact earlier
+          RefinedType(improvedBaseType, isExact, isNullable = false)
+      }
```

Comment:
It seems odd to do this here. Can't we do this at the call-site? (It seems all the information is there).

Reply:
At what call site, exactly? In `pretransformApply`? I tried that, and somehow it didn't go well. It was too far removed from the ultimate knowledge of what the `value` in the `Binding` is; I had to reproduce the decomposition of `PreTransBlock`s, etc. In the end, putting this here was the easiest, most DRY thing.